crates/core/tedge_agent/src/operation_file_cache/mod.rs

```diff
@@ -0,0 +1,360 @@
+        while let Some(message) = self.input_receiver.recv().await {
+            match message {
+                FileCacheInput::MqttMessage(message) => self.process_mqtt_message(message).await?,
+                FileCacheInput::IdDownloadResult(message) => self.process_download(message).await?,
+            }
+        }
+
+        Ok(())
+    }
+}
+
+impl FileCacheActor {
+    async fn process_mqtt_message(
+        &mut self,
+        mqtt_message: MqttMessage,
+    ) -> Result<(), RuntimeError> {
+        let Ok((entity, channel)) = self.mqtt_schema.entity_channel_of(&mqtt_message.topic) else {
+            return Ok(());
+        };
+
+        let Channel::Command {
+            operation: OperationType::ConfigUpdate,
+            cmd_id,
+        } = channel
+        else {
+            return Ok(());
+        };
+
+        let update_payload =
+            match serde_json::from_slice::<ConfigUpdateCmdPayload>(mqtt_message.payload.as_bytes())
+            {
+                Ok(payload) => payload,
+                Err(err) => {
+                    warn!("Received config update, but payload is malformed: {err}");
+                    return Ok(());
+                }
+            };
+
+        if update_payload.remote_url.is_empty() || update_payload.tedge_url.is_some() {
+            return Ok(());
+        }
+
+        match &update_payload.status {
+            CommandStatus::Executing => {
+                self.download_config_file_to_cache(&mqtt_message.topic, &update_payload)
+                    .await?;
+            }
+            CommandStatus::Successful => self.delete_symlink_for_config_update(
+                &entity,
+                &update_payload.config_type,
+                &cmd_id,
+            )?,
+            CommandStatus::Failed { .. } => self.delete_symlink_for_config_update(
+                &entity,
+                &update_payload.config_type,
+                &cmd_id,
+            )?,
+            _ => {}
+        }
+
+        Ok(())
+    }
+
+    async fn process_download(&mut self, download: IdDownloadResult) -> Result<(), RuntimeError> {
+        let (topic, result) = download;
+
+        let Some(mut operation) = self.pending_operations.remove(&topic) else {
+            return Ok(());
+        };
+
+        let (entity, Channel::Command { cmd_id, .. }) = self
+            .mqtt_schema
+            .entity_channel_of(&Topic::new(&topic).unwrap())
+            .expect("only topics targeting config update command should be inserted")
+        else {
+            return Ok(());
+        };
+
+        let download = match result {
+            // if cant download file, operation failed
+            Err(err) => {
+                let error_message = format!("tedge-agent failed downloading a file: {err}");
+                operation.failed(&error_message);
+                error!("{}", error_message);
+                let message = MqttMessage::new(
+                    &Topic::new(&topic).unwrap(),
```

Comment:
Minor. We have `Topic::new_unchecked()`, which suits here.

Reply:
fixed